src/main/scala/is/hail/expr/Type.scala

```diff
@@ -615,6 +672,44 @@ final case class TArray(elementType: Type) extends TIterable {
   override def genNonmissingValue: Gen[Annotation] =
     Gen.buildableOf[Array, Annotation](elementType.genValue).map(x => x: IndexedSeq[Annotation])
 
+  def ordering(missingGreatest: Boolean): Ordering[Annotation] =
+    annotationOrdering(extendOrderingToNull(missingGreatest)(
+      Ordering.Iterable(elementType.ordering(missingGreatest))))
+
+  override def unsafeOrdering(missingGreatest: Boolean): UnsafeOrdering = {
+    val eltOrd = elementType.unsafeOrdering(missingGreatest)
+    val mg = missingGreatest
+    val fund = fundamentalType
+
+    new UnsafeOrdering {
+      override def compare(r1: MemoryBuffer, o1: Int, r2: MemoryBuffer, o2: Int): Int = {
+        val length1 = loadLength(r1, o1)
+        val length2 = loadLength(r2, o2)
+
+        var i = 0
+        while (i < math.min(length1, length2)) {
+          val leftDefined = isElementDefined(r1, o1, i)
+          val rightDefined = isElementDefined(r2, o2, i)
+
+          if (leftDefined && rightDefined) {
```

Comment:
I'd `leftDefined.compare(rightDefined)` (adjusted by `mg`) and then return that if it is `!= 0` otherwise compare the element.